Comments: Enable user info and block user

Copons opened the pull request:
Close #13895

Enables the **User Info** panel and the **Block User** action.

The Block action simply adds (or removes) the comment author email address to the site's comment blacklist (as shown in _Settings → Discussion_).

### Screenshots

| Unblocked | Blocked |
| --- | --- |
| <img width="733" alt="screen shot 2017-09-25 at 12 21 47" src="https://user-images.githubusercontent.com/2070010/30806079-34327fca-a1ec-11e7-82ac-859f20001a2d.png"> | <img width="730" alt="screen shot 2017-09-25 at 12 22 04" src="https://user-images.githubusercontent.com/2070010/30806091-3cfc874a-a1ec-11e7-87d0-015ee045cf36.png"> |
| <img width="371" alt="screen shot 2017-09-20 at 17 50 01" src="https://user-images.githubusercontent.com/2070010/30656551-6ec69c52-9e2c-11e7-8fd2-85caec24b66d.png"> | <img width="394" alt="screen shot 2017-09-20 at 17 50 18" src="https://user-images.githubusercontent.com/2070010/30656561-73657774-9e2c-11e7-8292-b7b238a62007.png"> |

Bonus: anonymous user block message

<img width="729" alt="screen shot 2017-09-25 at 12 24 31" src="https://user-images.githubusercontent.com/2070010/30806169-8f655fde-a1ec-11e7-8b8a-5d5b0970cc12.png">

gwwar commented:
> I've not included any stats tracking because it's unclear to me if it's an useful action to track

💯 track all the things! Tracking user actions gives us a few things:
1. Historical baseline of how often people click on this. (sudden jumps/drop off are likely bugs or bad behavior)
2. We can slice the data later as we please

gwwar commented:
Should I be able to block myself 😆 ?

gwwar commented:
Looks like we have the UI option to block anon commenters, but this does nothing since no email address is available. Should we add the Anonymous name in this case, or related ip if available?

<img width="957" alt="screen shot 2017-09-20 at 3 23 01 pm" src="https://user-images.githubusercontent.com/1270189/30670579-b0cad696-9e17-11e7-9a0a-71659e7a35ab.png">

gwwar commented:
Great work so far @Copons!

rickybanister commented:
This seems totally great for now, but I had a thought for the future—perhaps we could put a little chevron next to 'Spam' in the actions panel to allow the user to block all future comments from that user vs just spamming a single one—basically a short cut. We might learn that asking users to expand the user info panel is an extra step (something to find out from real-world use, it's just a hunch).

folletto commented:
First of all: AMAZING, as I was thinking of suggesting exactly this, and there it is! Kerry already suggested it, and you're already doing it! 🎈 

> Should blocking a comment author also unapprove its comment?

I agree with you both. The user intent when blocking originates from the message, so it should be un-approved too. I back Kerry advice to double check first as it might be already a consequence of blocking, that we have to show in the UI?

> Should we add an Undo button in the block/unblock notice?

No, as the inverse option is right there under the mouse/finger.

> There are currently no visual clues to show that a comment author is blocked, except the highlighted Block icon. 

This is a bit tricky, as a "blocked" user shouldn't show up anymore in the comments queue. That said, I'd add a label near the name ("Blocked") and I'd replace the avatar with `gridicons-block` colored gray, matching the circle to have the same size.

Just to be clear: does the message disappear from the list once the user is blocked? 

Reason: the avatar might be a reason why the person was blocked (i.e. think of swastica or dickpic as avatar). We shouldn't ever show the avatar of a blocked person unless explicitly requested by the user (i.e. by visiting their site or analogous operation).


> I've not included any stats tracking because it's unclear to me if it's an useful action to track

It's a **MASSIVELY** useful thing to track, especially in this scenario as it would make this kind of data WAY more easy to access. :) 
In general as Kerry said: when in doubt, track. ;) 

> This seems totally great for now, but I had a thought for the future

I think in the future we need a more thorough way to deal with different forms of abuse, so I expect a more explicit interface, guided, with messages to support the individual and us to track the abuse types better. :) 

---

Extra input in terms of design:

1. Let's avoid border-less buttons when possible as they depend more on context to be understood as clickable. In this case let's use a standard white button.
2. In case of Anonymous comments, let's not have a button but link `/settings/discussion/` with a message along the lines of: "Anonymous messages can't be blocked individually, but you can update your [settings](https://wordpress.com/settings/discussion/) to allow comments only from registered users". This should be more effective long term, and it would make a more grounded advice in how to deal with anonymous abuse.
3. One shouldn't be able to block themselves. :P

Copons commented:
Thank you all for chiming in, super appreciated!

> The user intent when blocking originates from the message, so it should be un-approved too.

Of this I'm not so sure, you know? I might want to block a user but keep its comment.
Also, the behaviour is different than what happens on FB, for example, where blocking an user results in making them completely disappear.
On WP, instead, the action basically mean "put the email address in blacklist" and filter out all the future comments originating from that email address.
So, the true meaning is "block all future comments from this user".

Unapproving the comment might lead to a follow up question like "why all previous comments by this user are still approved?", which is a kind of annoying question for us to answer. 😄 

> [unapproving] might be already a consequence of blocking

Blocking doesn't have any consequences on whatever is already stored in the database.
Blocking is literally "add an email address into the blacklist".
And there is no "block user" on wp-admin, btw, so this is all new stuff.

This, of course, unless I've totally missed it in all these months working on comments. 😆 

> Just to be clear: does the message disappear from the list once the user is blocked?

To reiterate: nothing happens when a user is blocked: the blacklist filtering will be applied to the next submitted comment.

Though, with my current implementation, we keep the blacklist handy and check if the comment's author's email appears in it, and pass along the prop `userIsBlocked`.

With this prop, we can hide the avatar if we want, but I'd recommend doing so in another PR, because the author avatar appears in several different places, and it would increase this PR size by quite a lot.

Copons commented:
Updates:

- Added stats tracking for `block_user` and `unblock_user` actions.

- Replaced Block User link with a proper `Button`.

- Removed Block User from own comments.

- If the comment author doesn't have an email address (the author is anonymous), replaced Block User with directions to the site settings, to disable anonymous comments.

kwight commented on client/blocks/comment-detail/comment-detail-author.jsx:
```diff
@@ -71,9 +123,14 @@ export class CommentDetailAuthor extends Component {
 			translate,
 		} = this.props;
+
+		const showBlockUser = canUserBlacklist && !! authorEmail && ( authorEmail === currentUserEmail );
```
I think this supposed to be `( authorEmail !== currentUserEmail )`?

Copons replied:
Ah shoot, fumbled when moving things around, thanks for catching it!

kwight commented:
Oh, I see, you're ahead of me – we _do_ have that discussions blacklist already implemented in `settings/discussion`.

Copons commented:
Here you go @megsfulton!

<img width="725" alt="screen shot 2017-09-22 at 12 12 47" src="https://user-images.githubusercontent.com/2070010/30742099-91be2022-9f8f-11e7-8cb7-1e1c4c839946.png">

a8ci18n commented on client/blocks/comment-detail/comment-detail-author.jsx:
```diff
@@ -95,38 +152,50 @@ export class CommentDetailAuthor extends Component {
 						<Gridicon icon="globe" />
 						<span>
-							{ authorIp }
+							{ authorIp || <em>{ translate( 'No IP address' ) }</em> }
```
Hi! I've found a possible matching string that has already been translated **1** time:
`translate( 'IP address', 'NULL', { context: 'NULL'} )` *ES Score: 11*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/comments/user-info)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/blocks/comment-detail/comment-detail-author.jsx:
```diff
@@ -95,38 +152,50 @@ export class CommentDetailAuthor extends Component {
+							</ExternalLink>
+						}
+						{ ! authorUrl &&
+							<em>{ translate( 'No website' ) }</em>
```
Hi! I've found a possible matching string that has already been translated **106** times:
`translate( 'Website' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/blocks/comment-detail/comment-detail-author.jsx:
```diff
@@ -95,38 +152,50 @@ export class CommentDetailAuthor extends Component {
 						<Gridicon icon="mail" />
 						<span>
-							{ authorEmail }
+							{ authorEmail || <em>{ translate( 'No email address' ) }</em> }
```
Hi! I've found a possible matching string that has already been translated **1** time:
`translate( 'Your email address.', 'NULL', { context: 'NULL'} )` *ES Score: 9*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/comments/user-info)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

folletto commented:
> No IP address

Wasn't the IP address the only thing we _can_ actually show for anonymous users?

Minor details:

1. the text of "Anonymous messages can't..." doesn't seem having enough contrast. Maybe it's an artefact of the screenshot tho. Anyhow, let's double check it's using `$gray-text-min`, and... can we underline the hyperlink? (this is probably something we need to change somewhere else, but let's fix it here for the time being).
2. The same contrast thing should apply to the other fields, but we can keep the lighter gray when they are empty ("No email").
3. `un/block_user` actions: can we add a parameter that specifies if the block was on an anonymous user, a registered user, or an email-only user? This would help greatly to get a better sense of what's happening. Anything else you think we could anonymously track here useful for data analysis? 

Apart from this — it's great! I can't wait to get this merged. 
Small step forward in a way, but huge improvement from another perspective. :)

Copons commented:
@folletto 

Ah, yeah, the IP address should also be available, and in fact I had to modify that user manually to display the "No IP address". Though, having a fallback in place is always good, and in this case it's really minor.

1 & 2: good catch! I guess when Dave updated the grays, it didn't touch this section because it was hidden. Apparently _all_ the User Info block uses the lighter gray, and I guess _all_ of it should use the darker one instead, even the "no value" labels (that I've styled as italic btw).

3. Great suggestions!
And I honestly don't have any good ideas for other data to track.

folletto commented:
> Apparently all the User Info block uses the lighter gray, and I guess all of it should use the darker one instead, even the "no value" labels (that I've styled as italic btw).

👍

Copons commented:
Updates!

- Text contrast: text changed to `$gray-text-min` and gridicons to `$gray`.
<img width="584" alt="screen shot 2017-09-22 at 13 12 53" src="https://user-images.githubusercontent.com/2070010/30743959-cc55e9a6-9f97-11e7-9653-9b00a6f4ca4d.png">


- Stats: whole new thing.
  - Bumps: `calypso_comment_management` namespace with the following actions:
    - `comment_author_blocked`
    - `comment_author_unblocked`
  - Tracks: `calypso_comment_management_moderate_user` with the following properties:
    - `action = "block_user" | "unblock_user"`
    - `user_type = "wpcom" | "email_only" | "anonymous"`

Note that the `user_type: "anonymous"` is not currently supported since we block the email, and anons don't have it.
I've added it anyway as a possible user type in case we might want to block the IP address, for example.

folletto commented:
> Note that the user_type: "anonymous" is not currently supported since we block the email, and anons don't have it.

Mighty facepalm to me. Right. 
Can we track the number of people clicking the link then? :D  

Everything else: you're awesome! 🌟

Copons commented:
So, before I push, the change consists in:

- Bumps: `calypso_comment_management` namespace with the following actions:
  - `comment_author_blocked`
  - `comment_author_unblocked`
  - `open_discussion_settings` (in case of anon user)
- Tracks: `calypso_comment_management_moderate_user` with the following properties:
  - `action = "block_user" | "unblock_user" | "open_discussion_settings"`
  - `user_type = "wpcom" | "email_only" | "anonymous"`

I'm totally open to better naming suggestions though. 😄

folletto commented:
> Tracks

Labelling seems appropriate. I'm not 100% sure about adding "open_discussion_settings" under "moderate user" tho. Might be another item with a related name instead?

My suggestion for this is that I can easily in Tracks visualize the actions (name → Property: Action → View Trends) and I would get a nice overview all at once, but the link doesn't seem comparing apples to apples.

Copons commented:
> Labelling seems appropriate. I'm not 100% sure about adding "open_discussion_settings" under "moderate user" tho. Might be another item with a related name instead?
>
> My suggestion for this is that I can easily in Tracks visualize the actions (name → Property: Action → View Trends) and I would get a nice overview all at once, but the link doesn't seem comparing apples to apples.

I'm in doubt about that too, but my reasoning is that, even if the actions are different, the intent is the same.
If I click on that link is because:
- I want to disable anonymous comments.
- I want to moderate an anonymous user, by adding other non-email details to the blacklist.
- I want to do some stuff in my discussion settings.

Either way, the intent is to moderate, and therefore the event should be `_moderate_user`.
The property `action` is there to specify if the moderation happened directly (block/unblock) or discussion settings (maybe, because we won't be able to know if it then actually happens).

Copons commented:
FYI I've added the new tracking action, but it's not yet connected to the link's onClick.

gwwar commented on client/blocks/comment-detail/comment-detail-author.jsx:
```diff
@@ -59,9 +70,56 @@ export class CommentDetailAuthor extends Component {
 		gmtOffset( this.props.site )
 	).format( 'll LT' );
 
+	showMoreInfo = () => !! this.props.authorEmail || !! this.props.authorIp || !! this.props.authorUrl;
```
lodash `some` might work better here.

gwwar commented on client/blocks/comment-detail/comment-detail-author.jsx:
```diff
@@ -59,9 +70,56 @@ export class CommentDetailAuthor extends Component {
+		const analytics = {
+			action: authorIsBlocked ? 'unblock_user' : 'block_user',
+			user_type: authorId ? 'wpcom' : 'email_only',
+		};
```
I think it might be a bit clearer to keep the analytics logic in one place. Perhaps consider passing through `authorId` and `authorIsBlocked` to the updateBlacklist action.

Copons replied:
But then we'd need to move the logic into `mapDispatchToProps`, making it a bit harder to follow imho.

gwwar commented on client/blocks/comment-detail/comment-detail-author.jsx:
```diff
@@ -59,9 +70,56 @@ export class CommentDetailAuthor extends Component {
+
+		const analytics = {
+			action: authorIsBlocked ? 'unblock_user' : 'block_user',
+			user_type: authorId ? 'wpcom' : 'email_only',
```
It might also be interesting to pass through the userId here

Copons replied:
I'm a bit wary of logging user info, though.

Also, let's say we log the ID of registered users. What about `email_only`? Should we log the email?
I mean, this will be valuable info if we wanted to investigate users that get blocked often, but there's also a privacy concern to think of.

gwwar commented on client/blocks/comment-detail/comment-detail-author.jsx:
```diff
@@ -26,18 +37,18 @@ export class CommentDetailAuthor extends Component {
 		authorEmail: PropTypes.oneOfType( [ PropTypes.bool, PropTypes.string ] ),
 		authorId: PropTypes.number,
```
Are we sure we're passing this through? I don't believe this is wired up in CommentDetailComment.

Since CommentDetailAuthor is now connected, maybe consider allowing it to pull in it's own data deps, or passing through click handlers. So we can keep this as a block or a simple UI component.

Copons replied:
Oh shoot, I guess it was a leftover from the past, and I just assumed we used it somewhere.

gwwar replied:
Note that a larger refactor doesn't need to be done in this PR, I'm just noting that it can be cleaned up later.

gwwar commented on client/blocks/comment-detail/style.scss:
```diff
@@ -496,11 +499,19 @@ a.comment-detail__author-more-element {
 	}
+
+	strong {
+		float: none;
```
What does this do?

Copons replied:
The author block usually contains the author name and the author url, on the same line.
To work properly on `rtl` we added a `float: left` on the name (which is the `strong` there), that gets reversed as `float: right` on `rtl`.

Though, in the User Info block, the author block contains the author display name and the author username (when provided) on **two lines**.
The `float` basically places the display name and the username in the same position, and the former covers the latter.

gwwar commented on client/blocks/comment-detail/style.scss:
```diff
@@ -496,11 +499,19 @@ a.comment-detail__author-more-element {
+	}
+}
+
+.comment-detail.card .comment-detail__author-more-element-block-user .gridicon {
```
At this depth, I'd recommend a classname on the gridicon itself.

vindl commented:
Nice work @Copons! 

The only thing that I started noticing while testing that bothers me is not related to this PR. I'm referring to the duplication of author's gravatar and name upon expanding the info, which makes things look crowded:
<img src="https://user-images.githubusercontent.com/1182160/30799155-32f968f4-a1dc-11e7-940b-575414c187de.png" width="400px" />

This looks redundant to me and if you all agree I can spin up a separate PR to remove it.

Copons commented:
@vindl I agree with you as well. Though, this isn't hidden behind feature flag, so I'd rather fix it before merging. :)

@folletto / @megsfulton do you have a suggestion for a [gridicon](https://wpcalypso.wordpress.com/devdocs/design/gridicons) that would fit there? Maybe `user` or `user-circle`?

folletto commented:
Yep, `user-circle` is ideal there :)
